PyRDF/backend/Dist.py

```diff
@@ -226,55 +255,93 @@ def _get_clustered_ranges(self, nentries, treename, filelist,
-        remainder = numclusters % self.npartitions
+        logger.debug("{} clusters will be split along {} partitions.".format(
+            numclusters, self.npartitions))
 
-        i = 0  # Iterator
-        ranges = []
-        entries_to_process = 0
-
-        while i < numclusters:
-            index_start = i
-            start = clusters[i][0]
-            i = i + partSize
-            if remainder > 0:
-                i += 1
-                remainder -= 1
-            index_end = i
-            if i == numclusters:
-                end = clusters[-1][1]
-            else:
-                end = clusters[i - 1][1]
-
-            range_files = []
-            for idx in range(index_start, index_end):
-                current_file = clusters[idx][3]
-                if range_files and range_files[-1] == current_file:
-                    continue
-                range_files.append(clusters[idx][3])
+        """
+        This list comprehension builds the elements of the ranged tuples:
+
+        1. ``start``: The minimum entry among all the clusters considered in a
+           given partition. The offset of the current file is subtracted
+        2. ``end``: The maximum entry among all the clusters considered in a
+           given partition. The offset of the current file is subtracted
+        3. ``filelist``: The list of files that are span between entries
+           ``start`` and ``end``::
+
+                Filelist: [file_1,file_2,file_3,file_4]
+
+                Clustered range: [0,150]
+                file_1 holds entries [0, 100]
+                file_2 holds entries [101, 200]
+                Then the clustered range should open [file_1, file_2]
+
+                Clustered range: [150,350]
+                file_3 holds entris [201, 300]
+                file_4 holds entries [301, 400]
+                Then the clustered range should open [file_2, file_3, file_4]
+
+           Each ``cluster`` namedtuple has a ``fileandindex`` namedtuple. The
+           second element of this tuple corresponds to the index of the file in
+           the input `TChain`. This way all files can be uniquely identified,
+           even if there is some repetition (e.g. when building a TChain with
+           multiple instances of the same file). The algorithm to retrieve the
+           correct files for each range takes the unique filenames from the list
+           of clusters and sorts them by their index to keep the original order.
+
+           In each file only the clusters needed to process the clustered range
+           will be read.
+        4. ``friend_info``: Information about friend trees.
+        """
+        clustered_ranges = [
+            Range(
+                min(clusters)[0] - clusters[0].offset,  # type: int
+                max(clusters)[1] - clusters[0].offset,  # type: int
+                [
+                    filetuple.filename
+                    for filetuple in sorted(set([
```

Comment:
You need this to be a a set because `cluster.filetuple` can have repeated values?

Reply:
Exactly, the `clusters` list that is being iterated over in this list comprehension will look like
```
[
(0,100,0, (filename_1, 1)),
(100,200,0, (filename_1, 1)),
....,
(10000, 10100, 10000, (filename_2, 2)),
(10100, 10200, 10000, (filename_2,2)),
]
```
So we want filename_1 to appear only once to build the list of file names that the range will have to read.